Enable Avro message production.

derrickoswald opened the pull request:
The command `franz produce [topic] [flags]` previously only supported topics with String serialization (JSON).
This submission adds the ability to produce to topics with Avro serialization.
It adds the `--encode` flag to the produce command taking a schema name as parameter, which must be a name returned by `franz registry list` (usually topic name with '-value' suffix).

### Test
Set up a local [Confluent Platform](https://docs.confluent.io/platform/current/quickstart/ce-quickstart.html).
Install franz (go install).
Run `franz registry list` and confirm the schema `users-value` exists.
Run `franz produce pageviews --encode pageviews-value` and then on the subsequent line submit the JSON:
  **{"viewtime": 248888, "userid": "User_99", "pageid": "Page_99"}**
Confirm in [Control Center](http://localhost:9021/) that the record is added to the pageviews stream which is serialized with Avro.

### Files Affected
  cmd/produce.go - alter the init function to send binary data if the --encode flag is supplied with the schema name
  pkg/franz/producer.go - add a method to send binary data, and a codec member to support it
  README.md - update contributors

Commit:
Enable Avro message production.

The command "franz produce [topic] [flags]" previously only supported topics with String serialization (JSON).
This submission adds the ability to produce to topics with Avro serialization.
It adds the --encode flag to the produce command taking a schema name as parameter, which must be a name returned by "franz registry list" (usually topic name with '-value' suffix).

Test:
Set up a local Confluent Platform as described by https://docs.confluent.io/platform/current/quickstart/ce-quickstart.html.
Install franz (go install).
Run "franz registry list" and confirm the schema "users-value" exists.
Run "franz produce pageviews --encode pageviews-value" and then on the subsequent line submit the JSON:
  {"viewtime": 248888, "userid": "User_99", "pageid": "Page_99"}
Confirm in Control Center (http://localhost:9021/) that the record is added to the pageviews stream which is serialized with Avro.

Files affected:
  cmd/produce.go - alter the init function to encode and send binary data if the --encode flag is supplied with the schema name
  pkg/franz/franz.go - expose the codec member of the Franz structure.
  pkg/franz/producer.go - add a method to send binary data
  .gitignore - ignore IntelliJ files
  README.md - update contributors

derrickoswald commented:
For completeness, the franz config file ~/.config/config.yaml:
```
---

brokers:
  - localhost:9092

#tls:
#  key: path_to_key.pem
#  cert: path_to_cert.pem
#  caCert: path_to_chain.pem

registry: localhost:8081

kafka_version: 6.1.1
```

elwin commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+# IDEA #
+*.iml
+.idea
+*~
```
I wouldn't put that into a repositories .gitignore as this isn't strictly project-relevant and depends more on individual contributors - personally, I use a global ~/.gitignore that contains those sort of ignores.
However, definitely no harm done in adding them :)

elwin commented on README.md:
```diff
@@ -80,6 +80,9 @@ this project:
 - [Anastasios Zouzias](https://github.com/zouzias)
 - [Florian Lehner](https://github.com/florianl)
+- [Julien Perrochet](https://github.com/Shastick)
+- [Elvin Stephan](https://github.com/elwin)
```
```suggestion
- [Elwin Stephan](https://github.com/elwin)
```
😉

elwin commented on pkg/franz/franz.go:
```diff
@@ -98,6 +98,10 @@ func (f *Franz) Registry() Registry {
 	return f.registry
 }
 
+func (f *Franz) Codec() avroCodec {
```
This is an exported method with an unexported return-type which is a bit awkward as consumers cannot use this type (e.g. for method signatures or so).

Here I see two possible straightforward solutions:
1. Export `avroCodec`
2. Avoid exporting the codec all together and adjust `SendMessageBytes` to e.g. `SendMessageEncoded` that takes the same arguments as `SendMessage`, in addition to the schema ID - in there, the schema lookup and encoding would happen.

Personally, I'd prefer the latter approach as this keeps the logic a bit more encapsulated but both approaches have their justification.

elwin commented on cmd/produce.go:
```diff
@@ -30,6 +33,15 @@ Press Ctrl+D to exit.`,
 				}
 				defer producer.Close()
 
+				var schemaID uint32 = 0
```
No need to explicitly set to 0
```suggestion
				var schemaID uint32
```

elwin commented on cmd/produce.go:
```diff
@@ -30,6 +33,15 @@ Press Ctrl+D to exit.`,
 				defer producer.Close()
 
+				var schemaID uint32 = 0
+				if "" != encode {
```
```suggestion
				if encode != "" {
```

Would keep it consistent with the other comparisons

elwin commented on cmd/produce.go:
```diff
@@ -41,15 +53,27 @@ Press Ctrl+D to exit.`,
 
-					if err := producer.SendMessage(topic, line, key); err != nil {
-						return "", err
+					if "" != encode {
```
```suggestion
					if encode != "" {
```

elwin reviewed (commented):
Hi Derrick!
Awesome, I think this feature was also on the bucket list, glad to see it being added :D
Just a few minor nitpicks but generally looks good to me :)

Pinging @symaras as I think he's maintaining this project (I'm just a former intern who happens to be notified in this project :)

Commit:
Update per suggestions.

   Move the .gitignore file to ~/.config/git/ignore.
   Fix contributors typo.
   Move encoding into the Producer.
   Remove extraneous Codec function from Franz structure.

Shastick commented:
Good morning, thanks for the PR, and good to see you around @elwin :)

@derrickoswald, as a member of our org you can actually contribute to the internal repository: PRs are meant to be used by external contributors.

Commit:
Fix GOLINT whining.

elwin reviewed (approved):
@derrickoswald awesome, thanks!

@Shastick if we can't meet in real life at least we can on PRs 😄

derrickoswald commented:
Moved to internal repository. It will eventually be merged in from there.
